feat: ADR for design of a License Credential (#177)
allinbits/cosmos-cash: docs/Reference/architecture/CONTRIBUTING.md

# CONTRIBUTING

## Installation

> Cosmos Cash diagrams are based on PlantUML. 

Reviewer: based on? or developed using?

PlantUML can be installed on your local system through your platforms' package manager:

MacOS, through [Homebrew](https://brew.sh/)

```
> brew install plantuml
```

Windows, through [Chocolatey](https://chocolatey.org/)

```
> choco install plantuml
```

Linux-based systems, through your distributions package manager

> ## Contributing diagrams

Reviewer: let's use our standard Title case style capitalization for all headings, I did not not all of them below, it's handy to use https://capitalizemytitle.com/

> The source code for all diagrams SHOULD be stored in `./src` folder AND have the `.puml` suffix. Please refer to the [PlanUML documentation](https://plantuml.com) for further details.

Reviewer: ```suggestion
The source code for all diagrams SHOULD be stored in `./src` folder AND have the `.puml` suffix. See the [PlanUML documentation](https://plantuml.com) for further details.
```

Reviewer: we don't want to use Please, we already are friendly and helpful, per https://developers.google.com/style/tone#politeness-and-use-of-please

## Themes

> Where possible all diagrams will be rendered to have a similar look and feel. To achieve this add the following to your diagram:

Reviewer: ```suggestion
Where possible all diagrams are rendered to have a similar look and feel. To achieve this look and feel, add the following theme to your diagram:
```

Reviewer: now that we're out of the ADR (requires future tense, per Cosmos SDK convention) let's use standards for present tense, per https://developers.google.com/style/tense

```
@startuml myDiagram
!theme tendermint from ../themes/
...

@enduml
```

## How to generate diagram images

There are [many ways to run PlantUML](https://plantuml.com/running) to generate images from the `.puml` files, but this project prefers the following: 

* PlantUML CLI
* Embed PlantUML diagrams using PlantUML's proxy service in Markdown
* Visual Studio Code Plugin - best for development

### PlantUML CLI

> There is [makefile](.makefile) that will build all `src/*.puml` files. To create SVG images just run `make (svg|png)` as needed. `make clean` will get a clean folder ready for new build.
> 
> This is the preferred and cleanest way to create all the architecture documentation.

Reviewer: ```suggestion
The [makefile](.makefile) builds all `src/*.puml` files. To create SVG images, run `make (svg|png)` as needed. Run `make clean` to get a clean folder ready for a new build.

Using PlantUML and makefile is the preferred and cleanest way to create all the architecture documentation.
```

### Proxy service

This method uses PlantUML's [proxy service](https://plantuml.com/server) to put generate images and embed into a Markdown page. 

> To use the proxy service integration simply use:

Reviewer: ```suggestion
To use the proxy service integration, use:
```

Reviewer: it's tempting when we think something is easy, but let's omit "simply" in all cases, per https://developers.google.com/style/tone?hl=en#some-things-to-avoid-where-possible

```
![cached image](http://www.plantuml.com/plantuml/proxy?src=https://raw.github.com/plantuml/plantuml-server/master/src/main/webapp/resource/test2diagrams.txt)
```
Or if caching is required:
```
![uncached image](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.github.com/plantuml/plantuml-server/master/src/main/webapp/resource/test2diagrams.txt)
```

### Visual Studio Code

> If you use [Visual Studio Code](https://code.visualstudio.com/) there is [PlantUML plugin](https://marketplace.visualstudio.com/items?itemName=jebbs.plantuml) available. 
> 
> From here, one can preview diagrams through the command palette (CTRL+SHIFT+P on Windows/Linux or COMMAND+SHIFT+P on MacOS)

Reviewer: ```suggestion
If you use [Visual Studio Code](https://code.visualstudio.com/), a [PlantUML plugin](https://marketplace.visualstudio.com/items?itemName=jebbs.plantuml) is available. 

From here, you can preview diagrams by using the command palette (CTRL+SHIFT+P on Windows/Linux or COMMAND+SHIFT+P on macOS).
```
